### PR TITLE
fix bugs from log filter feature

### PR DESCRIPTION
**Pull request by whutwhu**
Interface wipeField had bug if the filter word used not in the query params in URL, for example, we defined "password" as a filter, and it crashed when below URL wiped:
localhost:12270/password/admin?name=value

Another bug in this interface if filter word included in param value but not in param name, its followed param's value was wiped anyway, and may miss to wipe the real one.
For example, we defined "password" as a filter, for the below URL:
 localhost:12270/admin?name1=password&name2=value&password=123

The result should be:
localhost:12270/admin?name1=password&name2=value&password=XXX

Due to the bug, now the wiped result:
 localhost:12270/admin?name1=password&name2=XXXXX&password=123

This PR fix above bugs in wipeField.

### CODE REVIEW

**Comment by randall** [comment]
[approve ci]

File: proxy/logging/LogFilter.h [hunk 1]
```diff
@@ -401,37 +401,54 @@ wipeField(char **field, char *pattern, const char *uppercase_field)
...
-      const char *p2 = strstr(p1, "=");
-      if (p2) {
-        p2++;
-        memcpy(temp_text, p1, (p2 - p1));
-        temp_text += (p2 - p1);
-        const char *p3 = strstr(p2, "&");
-        if (p3) {
-          for (int i = 0; i < (p3 - p2); i++) {
-            temp_text[i] = 'X';
+    if (!p1) {
```
Comment (sudheerv) [inline proxy/logging/LogFilter.h]: Seems like this is not needed anymore?

File: proxy/logging/LogFilter.h [hunk 2]
```diff
@@ -401,37 +401,54 @@ wipeField(char **field, char *pattern, const char *uppercase_field)
...
+      return;
+    } else {
+      while(p1) {
```
Comment (sudheerv) [inline proxy/logging/LogFilter.h]: It seems like it'd probably be cleaner and more readable, if we move the code in the while loop into a small utility function (as it has grown too long now :) )

**Comment by sudheerv** [comment]
@whutwhu Welcome to the community! Good work, added a couple of coding style comments.

File: proxy/logging/LogFilter.h (hunk 2 above)
Comment (SolidWallOfCode) [inline proxy/logging/LogFilter.h]: Why the `if`? The `while` loop won't execute if `p1` is `nullptr`.

**Comment by SolidWallOfCode** [comment]
Heh, didn't notice @sudheerv commented on the same point.

File: proxy/logging/LogFilter.h (hunk 1 above)
Comment (whutwhu) [reply proxy/logging/LogFilter.h]: Fixed it!

File: proxy/logging/LogFilter.h (hunk 2 above)
Comment (whutwhu) [reply proxy/logging/LogFilter.h]: Fixed it.

File: proxy/logging/LogFilter.h [hunk 3]
```diff
@@ -382,6 +382,30 @@ LogFilterString::_checkCondition(OperatorFunction f, const char *field_value, si
   return retVal;
 }
 
+static const char *
+findPtternFromParamName(const char *lookup_query_param, const char *pattern)
+{
+  const char *p1 = strstr(lookup_query_param, pattern);
+  while (p1) {
+    // wipe pattern in param name, need search again if find pattern in param value
+    const char *p10 = strstr(p1, "=");
```
Comment (SolidWallOfCode) [inline proxy/logging/LogFilter.h]: `strchr('=')` instead?

File: proxy/logging/LogFilter.h [hunk 4]
```diff
@@ -382,6 +382,30 @@ LogFilterString::_checkCondition(OperatorFunction f, const char *field_value, si
   return retVal;
 }
 
+static const char *
+findPtternFromParamName(const char *lookup_query_param, const char *pattern)
```
Comment (SolidWallOfCode) [inline proxy/logging/LogFilter.h]: `findPttern...`?

**Comment by masaori335** [comment]
[approve ci]

**Comment by masaori335** [comment]
[approve ci autest]

**Comment by randall** [comment]
[approve ci autest]

**Comment by sudheerv** [comment]
updated PR looks good to me!

File: proxy/logging/LogFilter.h [hunk 5]
```diff
@@ -382,6 +382,61 @@ LogFilterString::_checkCondition(OperatorFunction f, const char *field_value, si
...
+    memcpy(temp_text, p1, (p2 - p1));
+    temp_text += (p2 - p1);
+    const char *p3 = strchr(p2, '&');
+    if (p3) {
+      for (int i = 0; i < (p3 - p2); i++)
+        temp_text[i] = 'X';
+      temp_text += (p3 - p2);
+      memcpy(temp_text, p3, ((buf_dest + strlen(buf_dest)) - p3));
+    } else {
+      for (int i = 0; i < ((buf_dest + strlen(buf_dest)) - p2); i++) {
```
Comment (zizhong) [inline proxy/logging/LogFilter.h]: a performance killer to have strlen() in the for loop condition.

File: proxy/logging/LogFilter.h [hunk 6]
```diff
@@ -382,6 +382,61 @@ LogFilterString::_checkCondition(OperatorFunction f, const char *field_value, si
...
+    // ensure p1 is in the param name now
+    break;
+  }
+  return p1;
+}
+
+static void
+updatePatternForFieldValue(char **field, const char *p1, int field_pos, char *buf_dest)
+{
+  char tmp_text[strlen(buf_dest) + 10];
```
Comment (zizhong) [inline proxy/logging/LogFilter.h]: Why 10?

File: proxy/logging/LogFilter.h (hunk 5 above)
Comment (sudheerv) [reply proxy/logging/LogFilter.h]: Good catch. I believe most compilers these days are smart enough to optimise this sort of usage already, so in practice it is probably not being evaluated each iteration anyway.

Regardless, +1 to fix it.

FWIW, this was an existing piece of code lying around forever (I blame myself :( ) and not something that @whutwhu specifically changed in this PR (other than just modularising it).

File: proxy/logging/LogFilter.h [hunk 7]
```diff
@@ -382,6 +382,61 @@ LogFilterString::_checkCondition(OperatorFunction f, const char *field_value, si
   return retVal;
 }
 
+static const char *
+findPatternFromParamName(const char *lookup_query_param, const char *pattern)
+{
+  const char *p1 = strstr(lookup_query_param, pattern);
+  while (p1) {
```
[the remark this answers is not in the document]
Comment (whutwhu) [reply proxy/logging/LogFilter.h]: The variable name was from the original code, fixed it.

File: proxy/logging/LogFilter.h [hunk 8]
```diff
@@ -382,6 +382,61 @@ LogFilterString::_checkCondition(OperatorFunction f, const char *field_value, si
...
+      p1 = strstr(p11, pattern);
+      continue;
+    }
+    // ensure p1 is in the param name now
+    break;
+  }
+  return p1;
+}
+
+static void
```
[the remark this answers is not in the document]
Comment (whutwhu) [reply proxy/logging/LogFilter.h]: Did it.

File: proxy/logging/LogFilter.h (hunk 5 above)
Comment (whutwhu) [reply proxy/logging/LogFilter.h]: I didn't notice it from original code, and thank you @zizhong for your careful review!

File: proxy/logging/LogFilter.h (hunk 6 above)
Comment (whutwhu) [reply proxy/logging/LogFilter.h]: I think it didn't make sense for "10", I change it to 1.

File: proxy/logging/LogFilter.h [hunk 9]
```diff
@@ -382,6 +382,61 @@ LogFilterString::_checkCondition(OperatorFunction f, const char *field_value, si
...
+updatePatternForFieldValue(char **field, const char *p1, int field_pos, char *buf_dest)
+{
+  char tmp_text[strlen(buf_dest) + 10];
+  char *temp_text = tmp_text;
```
[the remark this answers is not in the document]
Comment (whutwhu) [reply proxy/logging/LogFilter.h]: updated the variable name.

**Comment by randall** [comment]
[approve ci]

File: proxy/logging/LogFilter.h [hunk 10]
```diff
@@ -394,44 +458,17 @@ wipeField(char **field, char *pattern, const char *uppercase_field)
...
-        return;
-      }
-
-      tmp_text[strlen(buf_dest)] = '\0';
-      strcpy(*field, tmp_text);
+    const char *pattern_in_param_name = findPatternFromParamName(lookup_query_param, pattern);
+    if (pattern_in_param_name) {
+      int field_pos         = pattern_in_param_name - lookup_query_param;
+      pattern_in_param_name = query_param + field_pos;
+      updatePatternForFieldValue(field, pattern_in_param_name, field_pos, buf_dest);
```
Comment (zizhong) [inline proxy/logging/LogFilter.h]: The code doesn't seem to be capable to handle the case `http:///path?email=aaa&email=bbb&email=another_email`?

File: proxy/logging/LogFilter.h [hunk 11]
```diff
@@ -382,6 +382,70 @@ LogFilterString::_checkCondition(OperatorFunction f, const char *field_value, si
...
+      }
+    }
+  } else {
+    return;
+  }
+
+  buf_dest_to_field[buf_dest_len] = '\0';
+  strcpy(*field, buf_dest_to_field);
+}
+
```
Comment (zizhong) [inline proxy/logging/LogFilter.h]: A general comment, it's better to add tests for the PR. We do have `Catch` for unit test and `Autest` for integration test available now for ATS development.

Comment (whutwhu) [reply proxy/logging/LogFilter.h]: Add autest og-filter.test.py in tests/gold_tests/logging/

**Comment by zwoop** [comment]
Sigh, why did you merge this with 7 commits .. This should have been squashed down to one single commit before merging.

**Comment by zwoop** [comment]
IF you need / want this for v9.0.0, please make a PR against the v9.0.x branch, with one single commit.

**Comment by zizhong** [comment]
@zwoop Sorry about it. Should've been careful about multiple commits.

**Comment by zwoop** [comment]
Ping? Do we still want this for 9.0.0 ?

**Comment by whutwhu** [comment]
Yes, I'll make one commit into 9.0.0. Thanks!

**Comment by zwoop** [comment]
This has been cherry-picked into 9.0.0, via another PR. The milestone should stay 9.0.0 on this PR, since it's the originating PR.

**Pull request merged**